collection_mutation.cc

```diff
@@ -309,15 +308,35 @@ deserialize_collection_mutation(const abstract_type& type, bytes_view in) {
     }
     auto nr = read_simple<uint32_t>(in);
     ret.cells.reserve(nr);
-    for (uint32_t i = 0; i != nr; ++i) {
-        // FIXME: we could probably avoid the need for size
-        auto ksize = read_simple<uint32_t>(in);
-        auto key = read_simple_bytes(in, ksize);
-        auto vsize = read_simple<uint32_t>(in);
-        // value_comparator(), ugh
-        auto value = atomic_cell_view::from_bytes(ctype.value_comparator()->imr_state().type_info(), read_simple_bytes(in, vsize));
-        ret.cells.emplace_back(key, value);
-    }
+
+    visit(type, make_visitor(
+        [&] (const collection_type_impl& ctype) {
+            // value_comparator(), ugh
+            auto& type_info = ctype.value_comparator()->imr_state().type_info();
+            for (uint32_t i = 0; i != nr; ++i) {
+                // FIXME: we could probably avoid the need for size
+                auto ksize = read_simple<uint32_t>(in);
+                auto key = read_simple_bytes(in, ksize);
+                auto vsize = read_simple<uint32_t>(in);
+                auto value = atomic_cell_view::from_bytes(type_info, read_simple_bytes(in, vsize));
```

Comment:
How do we maintain compatibility with Cassandra here?

Reply:
By storing a non-frozen UDT as a collection of cells, where each cell's key is serialized as a 2-byte short. Then `write_collection` and `write_cell` in sstables/mc/writer.cc do the right thing.

Actually I made a mistake by using uint16_t instead of int16_t. Cassandra uses Java's short type for these cell keys. The difference would only show for values >= 2^16 though (we would correctly handle such long UDTs while Cassandra would start doing weird things).

Reply:
What happens when I alter the UDT and add/remove the same column 32k times? Do you run out of keys, even if the UDT has the just a handful of fields, or you do not?

Reply:
You cannot remove a UDT field.
For adding too many fields or creating a type with too many fields I'm gonna add a check.